src/osd/OSD.cc

```diff
@@ -4372,6 +4374,65 @@ int OSD::handle_pg_peering_evt(
   }
 }
 
+bool OSD::maybe_wait_for_max_pg(int primary, spg_t pgid)
+{
+  const auto max_pgs_per_osd =
+    (cct->_conf->get_val<int64_t>("mon_max_pg_per_osd") *
+     cct->_conf->get_val<double>("osd_max_pg_per_osd_hard_ratio"));
+  RWLock::RLocker l(pg_map_lock);
+  if (pg_map.size() < max_pgs_per_osd) {
+    return false;
+  }
+  if (primary == whoami) {
+    want_to_create_primary++;
+  } else {
+    want_to_create_replica.emplace(pgid.pgid, primary);
+  }
+  dout(5) << __func__ << " withhold creation of pg " << pgid
+	  << ": " << pg_map.size() << " >= "<< max_pgs_per_osd << dendl;
+  return true;
+}
+
+void OSD::resume_creating_pg()
+{
+  assert(pg_map_lock.is_locked());
+
+  const auto max_pgs_per_osd =
+    (cct->_conf->get_val<int64_t>("mon_max_pg_per_osd") *
+     cct->_conf->get_val<double>("osd_max_pg_per_osd_hard_ratio"));
+  unsigned spare_pgs = max_pgs_per_osd - pg_map.size();
+  assert(spare_pgs > 0);
+  if (want_to_create_primary > 0) {
+    if (monc->sub_want("osd_pg_creates", last_pg_create_epoch, 0)) {
+      dout(4) << __func__ << ": resolicit pg creates from mon since "
+	      << last_pg_create_epoch << dendl;
+      monc->renew_subs();
+    }
+    if (spare_pgs <= want_to_create_primary) {
+      want_to_create_primary = 0;
+      return;
+    }
+    spare_pgs -= want_to_create_primary;
+    want_to_create_primary = 0;
+  }
+  if (!want_to_create_replica.empty()) {
+    auto m = new MOSDPGTemp(osdmap->get_epoch());
+    for (auto pg = want_to_create_replica.cbegin();
+	 pg != want_to_create_replica.cend();) {
+      auto& pgid = pg->first;
+      auto primary = pg->second;
+      // OSDMap::pg_to_up_acting_osds() will turn to up set if pg_temp is
+      // empty. to re-trigger a peering, we have to twiddle the pg mapping a
+      // little bit, see PG::should_restart_peering().
+      m->pg_temp[pgid] = {primary};
```

Comment:
Can we make it {}?  My concern is that a 1x pool, or an unhealthy cluster, may already map to [primary] so that this doesn't trigger an interval change.  Setting pg_temp to [] explicitly, though, could be made so that OSDMonitor always removes the empty mapping on the next epoch.  That won't trigger an interval change if the mapping is already [], but in that case we'll obviously already see an interval change when it starts mapping to an OSD that might have to create it.

Reply:
that's what I tried at first. but as the comment put, `OSDMap::pg_to_up_acting_osds()` will use the upset if pg_temp is empty. so an empty pg_temp won't trigger an interval change. see `OSD::advance_pg()`.

Reply:
anyway, I will explore more alternatives.

Reply:
i see.

maybe we can pass in who the requester is (so we have another peer with the pg), and we either do [primary] or do [peer,primary] if we have to do trigger an interval change?